### ISSUE TITLE
4.10.1 & 4.10.2: Cannot publish artifact 'compileScala.mapping' (PATH) as it does not exist.

### ISSUE THREAD

**Issue by drujd** [gradle/gradle#6849]
```
project(':scala-for-tests-only') {
    apply plugin: 'java-library'
    apply plugin: 'scala'
    apply plugin: 'maven'

    repositories {
        mavenCentral()
    }

    //scala joint compilation for tests
    sourceSets.test.java.srcDirs.each { sourceSets.test.scala.srcDir it }
    sourceSets.test.java.srcDirs = []

    dependencies {
        compile 'org.scala-lang:scala-library:2.12.6'
    }
}
```

This project (which contains only Java main classes, but utilises Scala for joint compilation of Java and Scala test classes) fails on :scala-for-tests-only:install with
```
Execution failed for task ':test-classes-only:install'.
> Could not publish configuration 'archives'
   > Cannot publish artifact 'compileScala.mapping' (PATH\test-classes-only\build\tmp\scala\compilerAnalysis\compileScala.mapping) as it does not exist.
```

This happens only on 4.10.1 and 4.10.2. 4.10 and older versions work.

**Comment by big-guy** [comment]
Yep, I've confirmed this.

As a workaround, you can create a `src/main/scala/ScalaForTestsWorkaround.scala` with something like `class ScalaForTestsWorkaround` in it.

Is `test-classes-only` used for sharing test fixtures/base classes with other projects or does it build other things in `src/main`?

**Comment by drujd** [comment]
The problem occurs in both cases. If there are some non-test java (but not scala) classes, the easy fix is to also add ```
sourceSets.main.java.srcDirs.each { sourceSets.main.scala.srcDir it }
sourceSets.main.java.srcDirs = []
```
so the joint compiler is used for the java classes. If there are only test classes in the project, then you have to use the ugly workaround. :/

**Comment by big-guy** [comment]
If there are only test classes, you could also put everything in "main" if you are sharing those classes with other projects.  This is what we do in gradle/gradle for our testing only infrastructure projects.

**Comment by brosander** [comment]
Seeing this as well in modules that have all java source + scalatests

**Comment by drujd** [comment]
@brosander That can be fixed by
```
sourceSets.main.java.srcDirs.each { sourceSets.main.scala.srcDir it }
sourceSets.main.java.srcDirs = []

```

**Comment by ymwang1984** [comment]
Seeing this as well in some of our production Scala projects

**Comment by drujd** [comment]
@big-guy Moving tests to main is a pretty ugly workaround.

**Comment by zatekusen** [comment]
We seeing this error appeared randomly even in normal module as well. Haven't investigated further but it does look like that it happens when we just rerun the same build and some tasks are cached.

**Comment by drujd** [comment]
Yup. With 4.10 this works well, but there are other bugs...

**Comment by big-guy** [comment]
This will be released in 4.10.3 (hopefully today)

**Comment by drujd** [comment]
Is the fix present also in 5.0?

**Comment by big-guy** [comment]
Yes
